src/seraphis_wallet/encrypted_file.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "crypto/chacha.h"
+#include "crypto/crypto.h"
+#include "file_io_utils.h"
+#include "serialization/binary_archive.h"
+#include "serialization/containers.h"
+#include "serialization/crypto.h"
+#include "serialization/serialization.h"
+#include "serialization/string.h"
+#include "storages/portable_storage_template_helper.h"
+#include "string_coding.h"
+
+#include <type_traits>
+
+struct encrypted_file
+{
+    std::string encrypted_data;
+    crypto::chacha_iv iv;
+
+    BEGIN_SERIALIZE_OBJECT()
+    VERSION_FIELD(0)
+    FIELD(encrypted_data)
+    FIELD(iv)
+    END_SERIALIZE()
+};
+
+template <class T> bool read_encrypted_file(std::string path, const crypto::chacha_key &key, T &struct_out)
+{
+    std::string buf;
+    if (!epee::file_io_utils::load_file_to_string(path, buf))
+        return false;
+
+    encrypted_file file;
+
+    binary_archive<false> file_ar{epee::strspan<std::uint8_t>(buf)};
+    if (!::serialization::serialize(file_ar, file))
+        return false;
+
+    std::string decrypted_data;
+    decrypted_data.resize(file.encrypted_data.size());
+    crypto::chacha20(file.encrypted_data.data(), file.encrypted_data.size(), key, file.iv, &decrypted_data[0]);
+
+    binary_archive<false> ar{epee::strspan<std::uint8_t>(decrypted_data)};
+    if (!::serialization::serialize(ar, struct_out))
+        return false;
+
+    return true;
+}
+
+// NOTE: if this were c++20, Concepts and `require` could be used to make this one function
+template <class T> bool read_encrypted_file_json(std::string path, const crypto::chacha_key &key, T &struct_out)
+{
+    std::string buf;
+    if (!epee::file_io_utils::load_file_to_string(path, buf))
+        return false;
+
+    encrypted_file file;
+
+    binary_archive<false> file_ar{epee::strspan<std::uint8_t>(buf)};
+    if (!::serialization::serialize(file_ar, file))
+        return false;
+
+    std::string decrypted_data;
+    decrypted_data.resize(file.encrypted_data.size());
+    crypto::chacha20(file.encrypted_data.data(), file.encrypted_data.size(), key, file.iv, &decrypted_data[0]);
+
+    std::string decoded = epee::string_encoding::base64_decode(decrypted_data);
```

Comment:
Like @vtnerd I don't see any good argument for Base64 here, but maybe you had one in mind, or you copied this from somewhere where we could then check whether the conditions / the use case are really similar or not.

Reply:
I don't remember why I did that, I think the argument was in a meeting somewhere that it was easier to copy?
will change it back